logstash-core/src/main/java/org/logstash/util/ExponentialBackoff.java

```diff
@@ -0,0 +1,65 @@
+package org.logstash.util;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Random;
+
+public class ExponentialBackoff {
+    private final Long maxRetry;
+    private final int[]  BACKOFF_SCHEDULE_MS = {100, 200, 400, 800, 1_600, 3_200, 6_400, 12_800, 25_600, 51_200};
+    private final int BACKOFF_MAX_MS = 60_000;
+
+    private static final Logger logger = LogManager.getLogger(ExponentialBackoff.class);
+
+    public ExponentialBackoff(Long maxRetry) {
+        this.maxRetry = maxRetry;
+    }
+
+    public <T> T retry(CheckedSupplier<T> action) throws RetryException {
+        Long attempt = 0L;
+
+        do {
+            try {
+                attempt++;
+                return action.get();
+            } catch (Exception ex) {
+                logger.error("Backoff retry exception", ex);
+            }
+
+            if (isRetry(attempt)) {
+                try {
+                    int ms = backoffTime(attempt);
+                    logger.info("Retry({}) will execute in {} second", attempt, ms/1000.0);
+                    Thread.sleep(ms);
+                } catch (InterruptedException e) {
+                    throw new RetryException("Backoff retry aborted", e);
+                }
+            }
+        } while (isRetry(attempt));
+
+        throw new RetryException("Reach max retry");
+    }
+
+    private int backoffTime(Long attempt) {
+        return (attempt - 1 < BACKOFF_SCHEDULE_MS.length)?
+                BACKOFF_SCHEDULE_MS[attempt.intValue() - 1] + new Random().nextInt(1000) :
+                BACKOFF_MAX_MS;
+    }
+
+    private boolean isRetry(Long attempt) {
```

Comment:
`isRetry` is a name for a getter, and a getter shouldn't use params, maybe a more effective name could be `needRetry` or `hasMoreRetry`.